frmts/jpeg/jpgdataset.cpp

```diff
@@ -3410,6 +3421,172 @@ void JPGDatasetCommon::DecompressMask()
+
+        if (EQUAL(aosTokens[0], "JPEG"))
+        {
+            if (ppszDetailedFormat)
+                *ppszDetailedFormat = VSIStrdup(
+                    GDALGetCompressionFormatForJPEG(m_fpImage).c_str());
+
+            const auto nSavedPos = VSIFTellL(m_fpImage);
+            VSIFSeekL(m_fpImage, 0, SEEK_END);
+            auto nFileSize = VSIFTellL(m_fpImage);
+            if (nFileSize > std::numeric_limits<size_t>::max())
+                return CE_Failure;
+            if (nFileSize > 4)
+            {
+                VSIFSeekL(m_fpImage, nFileSize - 4, SEEK_SET);
+                // Detect zlib compress mask band at end of file
+                // and remove it if found
+                uint32_t nImageSize = 0;
+                VSIFReadL(&nImageSize, 4, 1, m_fpImage);
+                CPL_LSBPTR32(&nImageSize);
+                if (nImageSize > 2 && nImageSize >= nFileSize / 2 &&
+                    nImageSize < nFileSize - 4)
+                {
+                    VSIFSeekL(m_fpImage, nImageSize - 2, SEEK_SET);
+                    GByte abyTwoBytes[2];
+                    if (VSIFReadL(abyTwoBytes, 2, 1, m_fpImage) == 1 &&
+                        abyTwoBytes[0] == 0xFF && abyTwoBytes[1] == 0xD9)
+                    {
+                        nFileSize = nImageSize;
+                    }
+                }
+            }
+            auto nSize = static_cast<size_t>(nFileSize);
+            if (ppBuffer)
+            {
+                if (pnBufferSize == nullptr)
+                {
+                    VSIFSeekL(m_fpImage, nSavedPos, SEEK_SET);
+                    return CE_Failure;
+                }
+                bool bFreeOnError = false;
+                if (*ppBuffer)
+                {
+                    if (*pnBufferSize < nSize)
+                    {
+                        VSIFSeekL(m_fpImage, nSavedPos, SEEK_SET);
+                        return CE_Failure;
+                    }
+                }
+                else
+                {
+                    *ppBuffer = VSI_MALLOC_VERBOSE(nSize);
+                    if (*ppBuffer == nullptr)
+                    {
+                        VSIFSeekL(m_fpImage, nSavedPos, SEEK_SET);
+                        return CE_Failure;
+                    }
+                    bFreeOnError = true;
+                }
+                VSIFSeekL(m_fpImage, 0, SEEK_SET);
+                if (VSIFReadL(*ppBuffer, nSize, 1, m_fpImage) != 1)
+                {
+                    if (bFreeOnError)
+                    {
+                        VSIFree(*ppBuffer);
+                        *ppBuffer = nullptr;
+                    }
+                    VSIFSeekL(m_fpImage, nSavedPos, SEEK_SET);
+                    return CE_Failure;
+                }
+
+                constexpr GByte EXIF_SIGNATURE[] = {'E', 'x',  'i',
+                                                    'f', '\0', '\0'};
+                constexpr const char APP1_XMP_SIGNATURE[] =
```

Comment:
`const` may be redundant if `constexpr` is also set.

Reply:
done